services/venice-router/src/test/java/com/linkedin/venice/router/throttle/ReadRequestThrottlerTest.java

```diff
@@ -368,4 +378,87 @@ public void testDisableMaxCapacityProtection() {
       // expected
     }
   }
+
+  /**
+   * We used to get NPEs due to subscribing listeners prior to the end of the constructor. This is a regression test
+   * for that issue.
+   */
+  @Test
+  public void testRaceConditionBetweenConstructorAndListeners() {
+    AtomicBoolean listenerThrewException = new AtomicBoolean(false);
+
+    /** This {@link ReadOnlyStoreRepository} simply calls one of the listeners right away. */
+    ReadOnlyStoreRepository storeRepository = new ReadOnlyStoreRepository() {
+      @Override
+      public Store getStore(String storeName) {
+        return null;
+      }
+
+      @Override
+      public Store getStoreOrThrow(String storeName) throws VeniceNoStoreException {
+        return null;
+      }
+
+      @Override
+      public boolean hasStore(String storeName) {
+        return false;
+      }
+
+      @Override
+      public Store refreshOneStore(String storeName) {
+        return null;
+      }
+
+      @Override
+      public List<Store> getAllStores() {
+        return Collections.emptyList();
+      }
+
+      @Override
+      public long getTotalStoreReadQuota() {
+        return 0;
+      }
+
+      @Override
+      public void registerStoreDataChangedListener(StoreDataChangedListener listener) {
+        try {
+          listener.handleStoreDeleted("testStore");
+        } catch (Throwable t) {
+          LOGGER.error("Caught a throwable when calling the listener immediately after subscription", t);
+        }
+      }
+
+      @Override
+      public void unregisterStoreDataChangedListener(StoreDataChangedListener listener) {
+      }
+
+      @Override
+      public int getBatchGetLimit(String storeName) {
+        return 0;
+      }
+
+      @Override
+      public boolean isReadComputationEnabled(String storeName) {
+        return false;
+      }
+
+      @Override
+      public void refresh() {
+      }
+
+      @Override
+      public void clear() {
+      }
+    };
+
+    ReadRequestThrottler readRequestThrottler = new ReadRequestThrottler(
+        mock(ZkRoutersClusterManager.class),
+        storeRepository,
+        mock(AggRouterHttpRequestStats.class),
+        mock(VeniceRouterConfig.class));
+
+    assertFalse(listenerThrewException.get(), "The listener should not receive an exception!");
```

Comment:
Value of `listenerThrewException` is never updated. So this assertion will always succeed.

Reply:
Ah, good catch 😅 ... I meant to update it in the catch of the registration function...

Reply:
Fixed in the latest commit.